dos/src/main.rs

```diff
@@ -235,7 +239,170 @@ impl TransactionGenerator {
     }
 }
 
-const SEND_BATCH_MAX_SIZE: usize = 1 << 10;
+// Multithreading-related functions
+//
+// The most computationally expensive work is signing new transactions.
+// Here we generate them in `num_gen_threads` threads.
+//
+enum TransactionMsg {
+    Transaction(Vec<Vec<u8>>, u64),
+    Exit,
+}
+/// number of transactions in one batch for sendmmsg
+const SEND_BATCH_MAX_SIZE: usize = 1 << 14;
+
+fn create_sender_thread(
+    tx_receiver: Receiver<TransactionMsg>,
+    mut n_alive_threads: usize,
+    target: &SocketAddr,
+    tpu_use_quic: bool,
+) -> thread::JoinHandle<()> {
+    // ConnectionCache is used instead of client because it gives ~6% higher pps
+    let tpu_use_quic = UseQUIC::new(tpu_use_quic).expect("Failed to initialize QUIC flags");
+    let connection_cache = ConnectionCache::new(tpu_use_quic, DEFAULT_TPU_CONNECTION_POOL_SIZE);
+    let connection = connection_cache.get_connection(target);
+
+    let timer_receiver = tick(Duration::from_millis(SAMPLE_PERIOD_MS as u64));
+
+    let mut time_send_ns = 0;
+    let mut time_generate_ns = 0;
+
+    thread::Builder::new().name("Sender".to_string()).spawn(move || {
+        let mut count: usize = 0;
+        let mut error_count = 0;
+
+        loop {
+            select! {
+                recv(tx_receiver) -> msg => {
+                    match msg {
+                        Ok(TransactionMsg::Transaction(data, time)) => {
+                            let len = data.len();
+                            let mut measure_send_txs = Measure::start("measure_send_txs");
+                            let res = connection.send_wire_transaction_batch_async(data);
+
+                            measure_send_txs.stop();
+                            time_send_ns += measure_send_txs.as_ns();
+                            time_generate_ns += time;
+
+                            if res.is_err() {
+                                error_count += len;
+                            }
+                            count += len;
+                        }
+                        Ok(TransactionMsg::Exit) => {
+                            info!("Worker is done");
+                            n_alive_threads -= 1;
```

Comment:
sorry, how does this work? it decrements this local n_alive_threads? which is set to the number of worker threads, and then when it hits 0 it exits?

Reply:
It didn't work, some leftovers which I left, sorry. I've updated the code with the following mechanism:
* There are N Generator threads which generate batches of transactions (size is configurable now)
* There is 1 Sender thread which sends them
* They communicate using channel

The communication through the channel might be of two types:
* Generator puts batches to channel and sender reads them
* Sender can decide to stop the program by dropping receiver of the channel in two cases: all the transactions have been sent or timeout. This logic is used for tests, when running standalone number of iterations is 0.

Reply:
ok, let me take a look again